File: OpenTokFs.RequestTypes/OpenTokDialRequest.cs
using Newtonsoft.Json;
using System.Collections.Generic;

namespace OpenTokFs.RequestTypes {
    public class OpenTokDialRequest {
        [JsonProperty("sessionId")]
        public string SessionId { get; set; } = "";

        [JsonProperty("token")]
        public string Token { get; set; } = "";

        [JsonProperty("sip")]
        public SipInfo Sip { get; set; } = new SipInfo();

        public class SipInfo {
            [JsonProperty("uri")]
            public string Uri { get; set; } = "";

            [JsonProperty("from", NullValueHandling = NullValueHandling.Ignore)]
            public string From { get; set; } = null;

            [JsonProperty("headers")]
            public Dictionary<string, string> Headers { get; set; } = new Dictionary<string, string>();

            [JsonProperty("auth", NullValueHandling = NullValueHandling.Ignore)]
            public Credentials Auth { get; set; } = null;

            [JsonProperty("secure")]
            public bool Secure { get; set; } = false;
        }

        public class Credentials {
            [JsonProperty("username")]
            public string Username { get; set; } = "";

            [JsonProperty("password")]
            public string Password { get; set; } = "";
        }
    }
}

File: OpenTokFs/RequestTypes/OpenTokDialRequest.fs
namespace OpenTokFs.RequestTypes

open System.Collections.Generic
open Newtonsoft.Json

module OpenTokDialRequest =
    [<AllowNullLiteral>]
    type Credentials () =
        member val username = "" with get, set
        member val password = "" with get, set

    type SipInfo() =
        member val uri = "" with get, set

        [<JsonProperty(NullValueHandling = NullValueHandling.Ignore)>]
        member val from = null with get, set

        member val headers = Dictionary<string, string> () with get, set

        [<JsonProperty(NullValueHandling = NullValueHandling.Ignore)>]
        member val auth: Credentials = null with get, set

        member val secure = false with get, set

type OpenTokDialRequest() =
    member val sessionId = "" with get, set
    member val token = "" with get, set
    member val sip = OpenTokDialRequest.SipInfo () with get, set